If gcc is configured to generate i686 code or better by default (like 
when using the --with-arch=pentium3 configure option), then the __i686
macro will always be defined automatically and thus screw up the
compilation of some .S files.
http://bugs.gentoo.org/131108
http://sourceware.org/ml/libc-alpha/2006-04/msg00090.html

2006-04-25  Mike Frysinger  <vapier@gentoo.org>

	* sysdeps/i386/sysdep.h (__i686): Undefine.

---
 nptl/sysdeps/pthread/pt-initfini.c |    3 +++
 sysdeps/i386/sysdep.h              |    8 ++++++++
 2 files changed, 11 insertions(+)

Index: eglibc-2.10/nptl/sysdeps/pthread/pt-initfini.c
===================================================================
--- eglibc-2.10.orig/nptl/sysdeps/pthread/pt-initfini.c	2009-12-01 13:40:19.000000000 +0100
+++ eglibc-2.10/nptl/sysdeps/pthread/pt-initfini.c	2009-12-01 13:40:31.000000000 +0100
@@ -50,6 +50,9 @@
 asm ("\n#define __i686 __i686");
 asm ("\n#endif");
 
+/* Embed an #include to pull in asm settings. */
+asm ("\n#ifdef __i686__\n#include <sysdep.h>\n#endif");
+
 /* The initial common code ends here. */
 asm ("\n/*@HEADER_ENDS*/");
 
Index: eglibc-2.10/sysdeps/i386/sysdep.h
===================================================================
--- eglibc-2.10.orig/sysdeps/i386/sysdep.h	2009-12-01 13:40:19.000000000 +0100
+++ eglibc-2.10/sysdeps/i386/sysdep.h	2009-12-01 13:40:31.000000000 +0100
@@ -18,6 +18,14 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
+/*
+ * When building for i686 targets or better, gcc automatically defines
+ * '__i686' to '1' for us which causes trouble when using section names
+ * like '__i686.get_pc_thunk.reg'.  Since we check for __i686__ in the
+ * code, killing '__i686' shouldn't be a problem.
+ */
+#undef __i686
+
 #include <sysdeps/generic/sysdep.h>
 
 #ifdef	__ASSEMBLER__
